torch/utils/data/_typing.py

```diff
@@ -338,6 +344,59 @@ def __hash__(self):
         return hash((self.__name__, self.type))
 
 
+def hook_iterator(namespace, profile_name):
+
+    def context():
+        return torch.autograd.profiler.record_function(profile_name)
+
+    class IteratorDecorator:
+        '''Wrap the iterator return result by adding __next__'''
+        def __init__(self, iterator):
+            self.iterator = iterator
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            with context():
+                return next(self.iterator)
+
+    func = namespace['__iter__']
+
+    if inspect.isgeneratorfunction(func):
+        @functools.wraps(func)
+        def wrap_generator(*args, **kwargs):
+            gen = func(*args, **kwargs)
+            try:
+                with context():
+                    response = gen.send(None)
+                while True:
+                    request = yield response
+                    with context():
+                        response = gen.send(request)
+            except StopIteration as e:
+                return e.value
+
+        namespace['__iter__'] = wrap_generator
+    else:
+        if '__next__' in namespace:
+            next_func = namespace['__next__']
+
+            @functools.wraps(next_func)
+            def wrap_next(*args, **kwargs):
+                with context():
+                    return next_func(*args, **kwargs)
+
+            namespace['__next__'] = wrap_next
+        else:
+            # have the __iter__ but not __next__ like what _ChildDataPipe did.
```

Comment:
Let's say we have `demux` and `fork` both existing in the pipeline. We won't be able to get the profiler name as `_Demux` or `_Forker`. All we can have is a `_Child`. It would be great if we can use a different name for different child

Reply:
Yes, you are right and it is a little pity. 

From my investigation result, it seems there is no way to get the _Forker/_Demux as they are the data member of _ChildDataPipe. https://github.com/pytorch/pytorch/blob/803e88d418beb924a79b9427a8aae9086877628a/torch/utils/data/datapipes/iter/combining.py#L166 

Do you have any better solution?

Reply:
Understood.
IMO, there is a way to at least add `__next__` to `_Demux` or `_Forker` to achieve that. Then, I won't block your PR for this case.

Reply:
I am not sure the design principle of _Demux and _Forker and am curios why it doesn't add __next__.